### ISSUE TITLE
How to bundle CommonJs third-party modules that fail at runtime - a suggested solution

### ISSUE THREAD

**Issue by GITHuHo** [driftyco/ionic-app-scripts#131]
#### TLDR

We have come up with a rather generic and simple solution that effectively solves any kind of Rollup bundling problems in Ionic RC0. The proposed solution just requires a tiny extension to the `bundle.js`  app-scripts source.
#### Description of the problem:

We are developing an IoT app with Ionic 2 that uses 3rd-party libs `mqtt` and `uuid`. Both npm packages are available in CommonJs format only. They run fine when bundled with previous Ionic 2 beta versions (i.e. webpack and browserify). 

With RC0 and @ionic/app-scripts, Rollup will bundle these dependencies without errors, but our app crashes at runtime. This is due to some sub-sub dependencies of `mqtt` and `uuid`, namely `readable-stream` and `parser` (for mqtt) and `des.js` (for uuid). Rollup/rollup-plugin-commonjs cannot bundle these modules properly for various reasons such as circular dependencies etc (see rollup forum for a discussion).

In the last few days a lot of Ionic RC0 users complained about Rollup issues. The easy ones, i.e. Rollup export errors at bundle time, can usually be solved as suggested by the Ionic documentation, such as namedExports (see https://ionicframework.com/docs/v2/resources/app-scripts/).

However, there is no solution yet for modules that fail at runtime. The suggested "Script Tag Include" workaround is rather poor, and doesn't work on libraries that do not define global variables.
#### Suggested solution

Let ionic-app-scripts create a Rollup bundle in **CommonJs** format, specifying which "faulty" modules to exclude, then hand the generated bundle over to Browserify. Browserify will detect the missing modules and properly include them in a browser-compatible bundle.

Install `browserify` as dev dependency in your app project:

```
npm install browserify --save-dev
```

Create custom config entries for `ionic_rollup` and `ionic_rollup_after_hook`(new feature) in package.json:

```
  "config": {
    "ionic_rollup": "./scripts/rollup.config.js",
    "ionic_rollup_after_hook": "./scripts/rollup.afterhook.js"
  },
```

Create the script file `./scripts/rollup.config.js` as follows:

```
const origRollupConfig = require('@ionic/app-scripts/config/rollup.config');

/**
 * format: The format of the generated bundle
 *
 * use `cjs` if external modules are specified; otherwise use `iife`
 */
origRollupConfig.format = 'cjs';

/**
 * dest: the output filename for the bundle in the buildDir
 *
 * use `../../.tmp/main.cjs.js` if external modules are specified; otherwise use `main.js`
 *
 * The intermediate rollup bundle should not be created in the build directory
 * because that triggers undesired live reloading by `ionic serve`
 */
origRollupConfig.dest = '../../.tmp/main.cjs.js';

/**
* external: An array of modules (in commonjs format) that should remain 
* external to the bundle. After bundling, browserify is run in an after hook
* function on the generated rollup bundle to produce a valid browser bundle that 
* includes the external modules in a proper way. 
*
* If external modules are specified, use format `cjs` and dest `../../.tmp/main.cjs.js`,
* otherwise use format `iife` and dest `main.js`
*/
origRollupConfig.external = ['mqtt', 'uuid'];

module.exports = origRollupConfig;
```

Define an after-hook function in `./scripts/rollup.afterhook.js`. This function is called by app-scripts after having written the rollup bundle. It will take care to run Browserify if needed, producing the final browser bundle named `main.js`. Note that the source maps produced by rollup in dev mode are left unchanged. i.e. there are no source maps for the excluded libraries in the final `main.js.map`.

```
var path = require('path');
var fse = require('fs-extra');
var browserify = require('browserify');

function rollupAfterHook(context, options, rollupConfig) {
    return new Promise(function(resolve, reject) {
        var hasExternals = Array.isArray(rollupConfig.external) && rollupConfig.external.length > 0;

        if (!hasExternals || rollupConfig.format !== 'cjs') {
            resolve();
            return;
        }

        // Run browserify on the generated rollup bundle to produce a valid 
        // browser bundle that properly includes the external modules.
        var error = undefined;
        var targetDest = path.join(context.buildDir, 'main.js');

        browserify(rollupConfig.dest, {
            cache: {},
            packageCache: {},
            // never add inline source map
            debug: false
        }).bundle()
            .on("error", function(err) {
                error = err;
                this.emit("end");
            })
            .pipe(fse.createWriteStream(targetDest))
            .on("finish", function() {                  
                if (error) {
                    reject(error);
                    return;
                }

                // TODO If rollupConfig.sourceMap is true,
                // extend rollup generated source map (currently in main.cjs.js.map)
                // by source map for external modules and write the resulting 
                // map to 'main.js.map'.

                if (rollupConfig.sourceMap) {
                    try {
                        fse.move(
                            rollupConfig.dest + ".map", targetDest + ".map", 
                            { clobber: true }, 
                            function (err) {});
                    } catch (e) {}
                }

                try {
                    fse.removeSync(rollupConfig.dest);
                } catch (e) {}

                resolve();
            });
    });
}

exports.rollupAfterHook = rollupAfterHook;
```

Finally, add invocation of the rollup after hook function in `@ionic/app-scripts/dist/bundle.js` (see code in second `then` clause):

```
    return rollup(rollupConfig).then(function (bundle) {
        logger_1.Logger.debug("bundle.modules: " + bundle.modules.length);
        // set the module files used in this bundle
        // this reference can be used elsewhere in the build (sass)
        context.moduleFiles = bundle.modules.map(function (m) { return m.id; });
        // async cache all the module paths so we don't need
        // to always bundle to know which modules are used
        setModulePathsCache(context.moduleFiles);
        // cache our bundle for later use
        bundleCache = bundle;
        // clean up any references
        rollupConfig.cache = rollupConfig.onwarn = rollupConfig.plugins = null;
        // write the bundle
        return bundle.write(rollupConfig);
    })
    // NEW CODE INSERTED HERE
    .then(function () {
        var afterHook = process.env['npm_package_config_ionic_rollup_after_hook'];
        if (afterHook) {
            return require(path_1.resolve(context.rootDir, afterHook))
              .rollupAfterHook(context, options, rollupConfig);
        }
        return true;
    })
    // END OF NEW CODE
    .catch(function (err) {
        // ensure references are cleared up when there's an error
        bundleCache = rollupConfig.cache = rollupConfig.onwarn = rollupConfig.plugins = null;
        return Promise.reject(err);
    });
```

The solution works perfectly in both dev and prod mode, as well as watching via `ionic serve`. As the after-hook function is an integral part of the bundling step, subsequent steps such as minify and uglify operate as expected. That's the reason why simply adding the post-bundling step in the npm `build` script is not feasible.

We would greatly appreciate if the Ionic Team could add such a hook function in the next version of app-scripts. Maybe, the specific browserify hook function could even become part of app-scripts, so that users only need to define the `external` property on a custom rollup config. What do you think?

Thanks for taking your time to review my suggestion.
Hubertus

**Comment by danbucholtz** [comment]
Hi @GITHuHo,

Thanks for putting this together. We'll chat about it. I like the idea in general.

Thanks,
Dan

**Comment by fifafu** [comment]
Very nice solution, at least until Rollup can handle such dependencies (if ever).

**Comment by timowolf** [comment]
👍

**Comment by GITHuHo** [comment]
Hi Timo,

danke für Dein hooray emoji! Jetzt hab ich schon 3 ☺

LG Hubertus

Von: Timo Wolf [mailto:notifications@github.com]
Gesendet: Mittwoch, 12. Oktober 2016 14:25
An: driftyco/ionic-app-scripts
Cc: Hohl, Hubertus (CT RDA ITP MSC-DE); Mention
Betreff: Re: [driftyco/ionic-app-scripts] How to bundle CommonJs third-party modules that fail at runtime - a suggested solution (#131)


👍

**Comment by TomDemulierChevret** [comment]
Great suggestion.

**Comment by danbucholtz** [comment]
@GITHuHo,

We decided to add Webpack support which should allow for bundling more deps. For now, we don't intend to do this. Great idea though and maybe we will revisit in the future!

Thanks,
Dan
